Replace filesystem-based test discovery with compile-time registry approach

corentinlange opened the pull request:
Refactors tanu's test discovery mechanism, replacing runtime filesystem
scanning with a compile-time registry-based approach. This change enables true modular test
organization where binaries only discover tests that are explicitly imported.

--

Found a way to streamline the test discovery by leveraging Rust's modules directly - it
simplifies the codebase while unlocking some cool modular test organization features.

--

I wanted to use tanu directly in my crate as another binary, but I encountered issues using it outside of main.rs

yukinarit commented:
Hi @corentinlange!

Thank you so much for the PR! Shame! I did not know `module_path!` works that nicely. I spent many nights to figure out and implement filesystem-based approach :sweat_smile: 

I love your suggestion. Let me review and test it on my projects first.

corentinlange commented:
Ok I changed some more things, to be honest I only use my tanu branch directly in my project and I didn't test the whole.

I revisited the whole thing. Normally build should be okay, import for test are fine and I added a second_runner and third_runner as sub binary in the tanu-integration-tests to test it out.

We might discuss about the naming not really sure about what I implemented.

corentinlange commented:
Update the clippy error

corentinlange commented:
Eh didn't think the 2+2 would block clippy didn't push it directly

yukinarit commented:
@corentinlange okay, let me check :+1:
